src/main/java/com/minecolonies/coremod/util/InventoryUtils.java

```diff
@@ -294,603 +363,1248 @@ public static boolean hasToolLevel(final String tool, final ItemStack stack, fin
     /**
      * Verifies if an item has an appropriated grade.
      *
-     * @param item     the type of tool needed
-     * @param level    the type of tool needed
-     * @param hutLevel the worker's hut level
+     * @param itemStack    the type of tool needed
+     * @param toolLevel    the type of tool needed
+     * @param minimalLevel the worker's hut level
      * @return true if tool is acceptable
      */
-    public static boolean verifyToolLevel(final ItemStack item, int level, final int hutLevel)
+    public static boolean verifyToolLevel(@NotNull final ItemStack itemStack, int toolLevel, final int minimalLevel)
     {
-        if (item == null || hutLevel > FREE_TOOL_CHOICE_LEVEL)
+        if (isItemStackEmpty(itemStack) || minimalLevel > FREE_TOOL_CHOICE_LEVEL || minimalLevel >= toolLevel)
         {
             return true;
         }
-        else if (item.isItemEnchanted() && hutLevel <= EFFECT_TOOL_CHOICE_LEVEL)
+        else if (itemStack.isItemEnchanted() && minimalLevel <= EFFECT_TOOL_CHOICE_LEVEL)
         {
             return false;
         }
-        else if (hutLevel >= level)
-        {
-            return true;
-        }
 
         return false;
     }
 
     /**
-     * Looks for a pickaxe to mine a block of {@code minLevel}.
+     * Adapted from {@link net.minecraft.entity.player.InventoryPlayer#addItemStackToInventory(ItemStack)}.
      *
-     * @param entity   inventory to check in.
-     * @param minlevel the needed pickaxe level
-     * @return true if a pickaxe was found
+     * @param itemHandler {@link IItemHandler} to add itemstack to.
+     * @param itemStack   ItemStack to add.
+     * @return True if successful, otherwise false.
      */
-    public static boolean isPickaxeInTileEntity(TileEntityChest entity, final int minlevel)
+    public static boolean addItemStackToItemHandler(@NotNull final IItemHandler itemHandler, @Nullable ItemStack itemStack)
     {
-        return InventoryFunctions.matchFirstInInventoryWithInventory(
-          entity,
-          stack -> stack != null && Utils.checkIfPickaxeQualifies(
-            minlevel,
-            Utils.getMiningLevel(stack, Utils.PICKAXE)
-          ),
-          InventoryFunctions::doNothing
-        );
+        if (!isItemStackEmpty(itemStack))
+        {
+            int slot;
+
+            if (itemStack.isItemDamaged())
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') NullPointerException might be thrown as 'itemStack' is nullable here [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2259)